File: slate/notes.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass(slots=True)
class NoteBlock:
    id: str
    type: str  # e.g., "text", "heading", "image", "code"
    content: str
    metadata: dict[str, Any] = field(default_factory=dict)
    created_at: str = field(default_factory=_now_iso)
    updated_at: str = field(default_factory=_now_iso)


@dataclass(slots=True)
class Note:
    id: str
    user_id: str
    title: str
    blocks: list[NoteBlock] = field(default_factory=list)
    folder_id: str | None = None
    created_at: str = field(default_factory=_now_iso)
    updated_at: str = field(default_factory=_now_iso)


class NotesModule:
    def __init__(self, user_id: str) -> None:
        self.user_id = user_id
        self._notes: dict[str, Note] = {}

    def create_note(self, note_id: str, title: str) -> Note:
        if note_id in self._notes:
            raise ValueError(f"Note with id {note_id} already exists.")

        note = Note(id=note_id, user_id=self.user_id, title=title)
        self._notes[note_id] = note
        return note

    def get_note(self, note_id: str) -> Note | None:
        return self._notes.get(note_id)

    def add_block(self, note_id: str, block_id: str, block_type: str, content: str) -> NoteBlock:
        note = self.get_note(note_id)
        if not note:
            raise ValueError(f"Unknown note: {note_id}")

        block = NoteBlock(id=block_id, type=block_type, content=content)
        note.blocks.append(block)
        note.updated_at = _now_iso()
        return block

    def list_notes(self) -> list[Note]:
        return list(self._notes.values())

File: src/notes.js
/**
 * Notes module foundations for Phase 3.
 */

export class NoteBlock {
  /**
   * @param {{ id: string, type: string, content: string, metadata?: Object, createdAt?: string, updatedAt?: string }} input
   */
  constructor(input) {
    if (!input?.id || !input?.type) {
      throw new Error('NoteBlock requires both id and type.');
    }

    this.id = input.id;
    this.type = input.type;
    this.content = input.content ?? '';
    this.metadata = input.metadata ?? {};
    this.createdAt = input.createdAt ?? new Date().toISOString();
    this.updatedAt = input.updatedAt ?? new Date().toISOString();
  }

  toJSON() {
    return {
      id: this.id,
      type: this.type,
      content: this.content,
      metadata: this.metadata,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

export class Note {
  /**
   * @param {{ id: string, userId: string, title: string, blocks?: Array<NoteBlock>, folderId?: string, createdAt?: string, updatedAt?: string }} input
   */
  constructor(input) {
    if (!input?.id || !input?.userId || !input?.title) {
      throw new Error('Note requires id, userId, and title.');
    }

    this.id = input.id;
    this.userId = input.userId;
    this.title = input.title;
    this.blocks = (input.blocks ?? []).map((block) => new NoteBlock(block));
    this.folderId = input.folderId ?? null;
    this.createdAt = input.createdAt ?? new Date().toISOString();
    this.updatedAt = input.updatedAt ?? new Date().toISOString();
  }

  toJSON() {
    return {
      id: this.id,
      userId: this.userId,
      title: this.title,
      blocks: this.blocks.map((block) => block.toJSON()),
      folderId: this.folderId,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

export class NotesModule {
  constructor(userId) {
    this.userId = userId;
    /** @type {Map<string, Note>} */
    this.notes = new Map();
  }

  createNote(noteId, title) {
    if (this.notes.has(noteId)) {
      throw new Error(`Note with id "${noteId}" already exists.`);
    }

    const note = new Note({ id: noteId, userId: this.userId, title });
    this.notes.set(noteId, note);
    return note;
  }

  getNote(noteId) {
    return this.notes.get(noteId);
  }

  addBlock(noteId, blockId, type, content) {
    const note = this.getNote(noteId);
    if (!note) {
      throw new Error(`Unknown note: ${noteId}`);
    }

    const block = new NoteBlock({ id: blockId, type, content });
    note.blocks.push(block);
    note.updatedAt = new Date().toISOString();
    return block;
  }

  listNotes() {
    return [...this.notes.values()].sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
  }
}
